### PR TITLE
Introduce DeferredSignatureChecker and have SignatureExtractorClass subclass it

### LINKED ISSUE #21151: signrawtransactionwithwallet fails with signed non-wallet inputs with many witness stack elements
I'm seeing a very strange behavior with `signrawtransactionwithwallet` when I provide signed non-wallet inputs.
The following sample transaction (regtest) is a lightning HTLC transaction with added `bitcoind` wallet inputs/outputs to bump the fee:

```
020000000001027697c418d3c56138f1ea0469026d081705290bd0317ad9a133319fa7b4ac59b10200000000010000001cd5376d42e12bad24911ae73b13c40b530a18c3da34de8d16fcf356bf2e32180000000000fdffffff02bb080000000000002200209357506c5a4a2df7c685481aad0302a3ccfe44c4e41122e33818861a37e2e6cccdfbea020000000016001487c7624537496328484f590d04a805f0c2a213e30500483045022100aef712b8d806077872e894343f917473497841b6d859bb7cf20473597353e7d8022030c39a0dc43b742ad0b746e8968b763e3053ec1001bbae4f5368d15d20d1152683483045022100a71dba3a0ed7d005c5e99b3a3e603026e60cd2708e7590162892bae9391965ba0220179d15465950863401956d0d4bbad0f413ce465b111dc71fe511d9bbf4200d0f0120b27cf03ab42eb16a475e289df7878d8c36d9465ba6b4f22ce1d21ef217e4f3d88d76a914c333ee02e6633f9bce0281ae0dc09e29b1b2761f8763ac672103bcdf028d4e53f7fbddb817091a60118b1d2fb7ab8cbab4b14ebeb981450d6ff77c8201208763a91487520d9fc7939063f7244ce842e4a5d2507d93d188527c21036df69837ccd9b8b353bdd37a8c616e3ea094ffee4c9512fec902ae8250ffb9f952ae6775022701b175ac6851b275680000000000
{
  "txid": "d922b3e02d77e7c522fcf46a308996df764dc61ef708a5702dc3dce6c2f58882",
  "hash": "1f558b7fd77d1339c48772f6ef12a0bf6cdbc05e371c6871d885afc89e6a5e3c",
  "version": 2,
  "size": 492,
  "vsize": 248,
  "weight": 990,
  "locktime": 0,
  "vin": [
    {
      "txid": "b159acb4a79f3133a1d97a31d00b290517086d026904eaf13861c5d318c49776",
      "vout": 2,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "",
        "3045022100aef712b8d806077872e894343f917473497841b6d859bb7cf20473597353e7d8022030c39a0dc43b742ad0b746e8968b763e3053ec1001bbae4f5368d15d20d1152683",
        "3045022100a71dba3a0ed7d005c5e99b3a3e603026e60cd2708e7590162892bae9391965ba0220179d15465950863401956d0d4bbad0f413ce465b111dc71fe511d9bbf4200d0f01",
        "b27cf03ab42eb16a475e289df7878d8c36d9465ba6b4f22ce1d21ef217e4f3d8",
        "76a914c333ee02e6633f9bce0281ae0dc09e29b1b2761f8763ac672103bcdf028d4e53f7fbddb817091a60118b1d2fb7ab8cbab4b14ebeb981450d6ff77c8201208763a91487520d9fc7939063f7244ce842e4a5d2507d93d188527c21036df69837ccd9b8b353bdd37a8c616e3ea094ffee4c9512fec902ae8250ffb9f952ae6775022701b175ac6851b27568"
      ],
      "sequence": 1
    },
    {
      "txid": "18322ebf56f3fc168dde34dac3180a530bc4133be71a9124ad2be1426d37d51c",
      "vout": 0,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "sequence": 4294967293
    }
  ],
  "vout": [
    {
      "value": 0.00002235,
      "n": 0,
      "scriptPubKey": {
        "asm": "0 9357506c5a4a2df7c685481aad0302a3ccfe44c4e41122e33818861a37e2e6cc",
        "hex": "00209357506c5a4a2df7c685481aad0302a3ccfe44c4e41122e33818861a37e2e6cc",
        "reqSigs": 1,
        "type": "witness_v0_scripthash",
        "addresses": [
          "bcrt1qjdt4qmz6fgkl0359fqd26qcz50x0u3xyusgj9cecrzrp5dlzumxq6sxuyz"
        ]
      }
    },
    {
      "value": 0.48954317,
      "n": 1,
      "scriptPubKey": {
        "asm": "0 87c7624537496328484f590d04a805f0c2a213e3",
        "hex": "001487c7624537496328484f590d04a805f0c2a213e3",
        "reqSigs": 1,
        "type": "witness_v0_keyhash",
        "addresses": [
          "bcrt1qslrky3fhf93jsjz0tyxsf2q97rp2yylr3fd72n"
        ]
      }
    }
  ]
}
```

Notice that the first input (the non-wallet one) has a stack with 5 elements (the first of which is empty). It is correctly signed and passes script validation.

When I send this transaction to `signrawtransactionwithwallet`, `bitcoind` returns the following error: `Unable to sign input, invalid stack size (possibly missing key) (code: -1)` referencing the non-wallet input.

But it does correctly sign the wallet input and returns the following transaction:

```
020000000001027697c418d3c56138f1ea0469026d081705290bd0317ad9a133319fa7b4ac59b10200000000010000001cd5376d42e12bad24911ae73b13c40b530a18c3da34de8d16fcf356bf2e32180000000000fdffffff02bb080000000000002200209357506c5a4a2df7c685481aad0302a3ccfe44c4e41122e33818861a37e2e6cccdfbea020000000016001487c7624537496328484f590d04a805f0c2a213e3018d76a914c333ee02e6633f9bce0281ae0dc09e29b1b2761f8763ac672103bcdf028d4e53f7fbddb817091a60118b1d2fb7ab8cbab4b14ebeb981450d6ff77c8201208763a91487520d9fc7939063f7244ce842e4a5d2507d93d188527c21036df69837ccd9b8b353bdd37a8c616e3ea094ffee4c9512fec902ae8250ffb9f952ae6775022701b175ac6851b275680247304402201566b009b4e392ca16d0e5658e7fd2257e511e8974f198fb8ac265473d492949022001c9b41d261f8eacc19c41f179aa36a449200fe66abb8f3aab45c5d49e280f2901210271f8ed45eddce4cb9f990998f1d28f91fe4f747a7d67900b5382484276ae4bfd00000000
{
  "txid": "d922b3e02d77e7c522fcf46a308996df764dc61ef708a5702dc3dce6c2f58882",
  "hash": "06722d22306472337f1229ad9642c3530f161b7ef3d54e061acf7ca09fb8ed76",
  "version": 2,
  "size": 418,
  "vsize": 229,
  "weight": 916,
  "locktime": 0,
  "vin": [
    {
      "txid": "b159acb4a79f3133a1d97a31d00b290517086d026904eaf13861c5d318c49776",
      "vout": 2,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "76a914c333ee02e6633f9bce0281ae0dc09e29b1b2761f8763ac672103bcdf028d4e53f7fbddb817091a60118b1d2fb7ab8cbab4b14ebeb981450d6ff77c8201208763a91487520d9fc7939063f7244ce842e4a5d2507d93d188527c21036df69837ccd9b8b353bdd37a8c616e3ea094ffee4c9512fec902ae8250ffb9f952ae6775022701b175ac6851b27568"
      ],
      "sequence": 1
    },
    {
      "txid": "18322ebf56f3fc168dde34dac3180a530bc4133be71a9124ad2be1426d37d51c",
      "vout": 0,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "304402201566b009b4e392ca16d0e5658e7fd2257e511e8974f198fb8ac265473d492949022001c9b41d261f8eacc19c41f179aa36a449200fe66abb8f3aab45c5d49e280f2901",
        "0271f8ed45eddce4cb9f990998f1d28f91fe4f747a7d67900b5382484276ae4bfd"
      ],
      "sequence": 4294967293
    }
  ],
  "vout": [
    {
      "value": 0.00002235,
      "n": 0,
      "scriptPubKey": {
        "asm": "0 9357506c5a4a2df7c685481aad0302a3ccfe44c4e41122e33818861a37e2e6cc",
        "hex": "00209357506c5a4a2df7c685481aad0302a3ccfe44c4e41122e33818861a37e2e6cc",
        "reqSigs": 1,
        "type": "witness_v0_scripthash",
        "addresses": [
          "bcrt1qjdt4qmz6fgkl0359fqd26qcz50x0u3xyusgj9cecrzrp5dlzumxq6sxuyz"
        ]
      }
    },
    {
      "value": 0.48954317,
      "n": 1,
      "scriptPubKey": {
        "asm": "0 87c7624537496328484f590d04a805f0c2a213e3",
        "hex": "001487c7624537496328484f590d04a805f0c2a213e3",
        "reqSigs": 1,
        "type": "witness_v0_keyhash",
        "addresses": [
          "bcrt1qslrky3fhf93jsjz0tyxsf2q97rp2yylr3fd72n"
        ]
      }
    }
  ]
}
```

**Notice how it drops all witness stack elements from the non-wallet input except the last one**: this seems to be the culprit. If I ignore `bitcoind`'s error and restore the witness stack of the non-wallet input, the transaction becomes valid and can be correctly published.

I tested this with Bitcoin Core 0.20.1 and 0.21.0, let me know if you want me to test against other versions.

### PR DESCRIPTION
**Pull request by achow101**
Previously SignatureExtractorChecker took a MutableTransactionSignatureChecker and passed through function calls to that. However not all functions were implemented so not everything passed through as it should have. To solve this, SignatureExctractorChecker now implements all of those functions via a new class - DeferredSignatureChecker. DeferredSignatureChecker is introduced to allow for future signature checkers which use another SignatureChecker but need to be able to do somethings outside of just the signature checking.

Fixes #21151

### CODE REVIEW

File: src/script/sign.cpp [hunk 1]
```diff
@@ -267,6 +267,10 @@ class SignatureExtractorChecker final : public BaseSignatureChecker
         }
         return false;
     }
+    // These methods just pass through to checker
+    bool CheckSchnorrSignature(Span<const unsigned char> sig, Span<const unsigned char> pubkey, SigVersion sigversion, const ScriptExecutionData& execdata, ScriptError* serror = nullptr) const { return checker.CheckSchnorrSignature(sig, pubkey, sigversion, execdata, serror); }
```
Comment (MarcoFalke) [inline src/script/sign.cpp:271]:
```
/usr/bin/ccache g++ -m64 -std=c++17 -DHAVE_CONFIG_H -I. -I../src/config   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I. -I./secp256k1/include  -DBOOST_SP_USE_STD_ATOMIC -DBOOST_AC_USE_STD_ATOMIC -isystem /tmp/cirrus-ci-build/depends/x86_64-pc-linux-gnu/include -I./leveldb/include -I./leveldb/helpers/memenv -I./univalue/include -I/tmp/cirrus-ci-build/depends/x86_64-pc-linux-gnu/include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -DPROVIDE_MAIN_FUNCTION  -fstack-reuse=none -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection   -Werror=vla -Werror=switch -Werror=unused-variable -Werror=date-time -Werror=return-type -Werror=sign-compare -Werror=suggest-override   -fPIE -pipe -O2  -fno-extended-identifiers -c -o script/libbitcoin_common_a-sign.o `test -f 'script/sign.cpp' || echo './'`script/sign.cpp
script/sign.cpp:271:10: error: ‘virtual bool {anonymous}::SignatureExtractorChecker::CheckSchnorrSignature(Span<const unsigned char>, Span<const unsigned char>, SigVersion, const ScriptExecutionData&, ScriptError*) const’ can be marked override [-Werror=suggest-override]
  271 |     bool CheckSchnorrSignature(Span<const unsigned char> sig, Span<const unsigned char> pubkey, SigVersion sigversion, const ScriptExecutionData& execdata, ScriptError* serror = nullptr) const { return checker.CheckSchnorrSignature(sig, pubkey, sigversion, execdata, serror); }
      |          ^~~~~~~~~~~~~~~~~~~~~
cc1plus: some warnings being treated as errors

Comment (achow101) [reply src/script/sign.cpp]: Changed the approach entirely, shouldn't be a problem.

**Comment by jonatack** [comment]
Some test coverage might be good (if this solves the issue).

**Comment by achow101** [comment]
I've added test cases (one CSV, one CLTV) for this.

**Comment by sipa** [comment]
I think it would be a bit cleaner and more flexible to have a DeferringSignatureChecker or so in script/sign which just forwards all calls to another BaseSignatureChecker it holds a pointer to, and then have the SignatureExtractorChecker extend that, overriding just what it needs.

That has the advance of just being able to use the SignatureCreator's Checker() as object deferred to (which in #21365 is extended to hold PrecomputedTransactionData etc).

**Comment by achow101** [comment]
Implemented @sipa's suggestion

**Comment by sipa** [comment]
utACK a97a9298cea085858e1a65a5e9b20d7a9e0f7303

**Comment by MarcoFalke** [comment]
Marked for backport

**Review by meshcollider — COMMENTED** [review]
Code review ACK a97a9298cea085858e1a65a5e9b20d7a9e0f7303

Confirmed that the test fails on master but passes with 6965456c10c9c4025c71c5e24fa5b27b15e5933a

**Review by instagibbs — APPROVED** [review]
utACK a97a9298cea085858e1a65a5e9b20d7a9e0f7303

**Comment by achow101** [comment]
Backported in #21640

**Comment by curtcurt871** [comment]
Need help with making a wallet for input. Need help with code and dev. Just send video instructions please and thank u

**Pull request merged**